zmij.cc

```diff
@@ -757,28 +803,29 @@ auto write_significand(char* buffer, uint64_t value) noexcept -> char* {
   uint32_t abb = abbcc / 100;
   uint32_t cc = abbcc % 100;
   auto [a, bb] = divmod100(abb);
+  auto [dd, ee] = divmod100(ddee);
 
   *buffer = char('0' + a);
   buffer += a != 0;
-  write2digits(buffer + 0, bb);
-  write2digits(buffer + 2, cc);
-  buffer += 4;
 
-  if (ffgghhii == 0) {
-    if (ddee != 0) return write4digits_trim_zeros(buffer, ddee);
-    return buffer - num_trailing_zeros[cc] - (cc == 0) * num_trailing_zeros[bb];
+  // Use an intermediate uint64_t to make sure that the compiler constructs
+  // the value in a register.  This way the buffer is written to memory in
+  // one go and count_trailing_nonzeros doesn't have to load from memory.
+  uint64_t buf = write8digits(bb, cc, dd, ee);
+  memcpy(buffer, &buf, 8);
+
+  if (ffgghhii == 0) [[unlikely]] {
```

Comment:
Why unlikely?

Reply:
There are 10^8 possible combinations for the trailing 8 digits but only one of those combinations takes that path.

That is of course an approximate statement given that not every decimal is representable as binary, but I assume it's close enough.

The actual distribution is probably also skewed by user input which will tend shorter. But the case where the user prints their manual input is probably not the most performance-critical?

Reply:
I'm not sure we can make such assumptions about the distribution of user data. It is very much possible that all of it is rounded or something like that. In any case I don't think it has significant effect so I just removed the attribute for now.